lib/Runtime/Library/LazyJSONString.h

```diff
@@ -0,0 +1,89 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+
+#pragma once
+
+namespace Js
+{
+
+struct JSONProperty;
+struct JSONArray;
+
+enum class JSONContentType : uint8
+{
+    Array,
+    Object,
+    Undefined,
+    Null,
+    True,
+    False,
+    Number,
+    String
+};
+
+typedef DListCounted<JSONProperty, Recycler> JSONObject;
```

Comment:
Does this need to be a dlist? Or would an slist suffice? Best I can tell, we always just iterate these lists from the beginning?

Reply:
And actually, I wonder why JSONObject and JSONArray need to be separate types- can JSONObject just be represented as JSONArray- they both look like a list of JSONProperties + Count, and you might be able to avoid a bunch of intermediate allocations

---
In reply to: [147304974](https://github.com/Microsoft/ChakraCore/pull/4077#discussion_r147304974) [](ancestors = 147304974)

Reply:
I think it could be, but problem with our SLists is that we don't have any way to append. Maybe I'm missing something, but seems like you can only push and pop, which would cause the list to be reversed. I could use SList and then reverse it, but I thought that might be more expensive.

Reply:
I don't think I can know up front how many properties I am going to enumerate for Objects?

Reply:
Ah yeah, forgot append semantics were missing in slists. I wonder how expensive doing two passes with the static enumerator is- if its not too expensive, you might get nice locality benefits by precomputing the size and allocating an array instead of a list?

Reply:
Although if possible, I'd do it heuristically here- if the number of properties is small, allocate an array- if its large, use a list since it would likely be allocating in a bucket that is generally sparse and cause fragmentation